File: crates/knot-core/resources/python/lsp.py
"""Knot Python LSP Support"""

import pydoc
import builtins


def get_hover(topic):
    """Documentation from __main__ context."""
    try:
        import __main__
        main_dict = __main__.__dict__

        obj = main_dict.get(topic)
        if obj is None:
            try:
                obj = eval(topic, main_dict)
            except:
                pass

        if obj is not None:
            doc = pydoc.getdoc(obj)
            if not doc:
                doc = pydoc.render_doc(obj, renderer=pydoc.plaintext)
            return doc
        else:
            return pydoc.render_doc(topic, renderer=pydoc.plaintext)
    except Exception as e:
        return f"No help found for '{topic}'"


def get_completions(token):
    """Completions from __main__ context."""
    try:
        import __main__
        main_dict = __main__.__dict__

        if '.' in token:
            parts = token.split('.')
            base_name = parts[0]
            prefix = parts[-1]

            obj = main_dict.get(base_name)
            if obj is not None:
                for part in parts[1:-1]:
                    obj = getattr(obj, part, None)
                    if obj is None: break

            if obj is not None:
                return "\n".join([attr for attr in dir(obj) if attr.startswith(prefix) and not attr.startswith('_')])
        else:
            candidates = list(main_dict.keys()) + dir(builtins)
            return "\n".join([c for c in candidates if c.startswith(token) and not c.startswith('_')])
    except:
        return ""
    return ""

File: crates/knot-core/resources/r/lsp.R
# Knot R LSP Support

get_hover <- function(topic) {
  tryCatch({
    h <- utils::help(topic)
    if (length(h) > 0) {
      tf <- tempfile()
      tools::Rd2txt(utils:::.getHelpFile(h), out = tf, options = list(underline_titles = FALSE))
      content <- readLines(tf)
      unlink(tf)
      paste(content, collapse = "\n")
    } else {
      ""
    }
  }, error = function(e) {
    ""
  })
}

get_completions <- function(token) {
  tryCatch({
    if (grepl("\\$", token)) {
      parts <- strsplit(token, "\\$")[[1]]
      obj_name <- parts[1]
      prefix <- if (length(parts) > 1) parts[2] else ""

      if (exists(obj_name, envir = .GlobalEnv)) {
        obj <- get(obj_name, envir = .GlobalEnv)
        if (is.list(obj) || is.data.frame(obj)) {
          n <- names(obj)
          matches <- n[startsWith(n, prefix)]
          paste(matches, collapse = "\n")
        } else {
          ""
        }
      } else {
        ""
      }
    } else {
      # Global completions
      matches <- utils::apropos(sprintf("^%s", token))
      paste(matches, collapse = "\n")
    }
  }, error = function(e) {
    ""
  })
}
